sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -906,19 +946,123 @@ public void setValueDeserializer(String valueDeserializer) {
+      // "beam_fn_api_use_deprecated_read" is not enabled.
+      if (!ExperimentalOptions.hasExperiment(input.getPipeline().getOptions(), "beam_fn_api")
+          || ExperimentalOptions.hasExperiment(
+              input.getPipeline().getOptions(), "beam_fn_api_use_deprecated_read")) {
+        // Handles unbounded source to bounded conversion if maxNumRecords or maxReadTime is set.
+        Unbounded<KafkaRecord<K, V>> unbounded =
+            org.apache.beam.sdk.io.Read.from(
+                toBuilder().setKeyCoder(keyCoder).setValueCoder(valueCoder).build().makeSource());
+
+        PTransform<PBegin, PCollection<KafkaRecord<K, V>>> transform = unbounded;
+
+        if (getMaxNumRecords() < Long.MAX_VALUE || getMaxReadTime() != null) {
+          transform =
+              unbounded.withMaxReadTime(getMaxReadTime()).withMaxNumRecords(getMaxNumRecords());
+        }
+
+        return input.getPipeline().apply(transform);
+      } else {
+        // If extractOutputTimestampFn is not set, use processing time by default.
+        SerializableFunction<KafkaRecord<K, V>, Instant> timestampFn;
+        if (getExtractOutputTimestampFn() != null) {
+          timestampFn = getExtractOutputTimestampFn();
+        } else {
+          timestampFn = ReadViaSDF.ExtractOutputTimestampFns.useProcessingTime();
+        }
+        ReadViaSDF<K, V> readTransform =
+            ReadViaSDF.<K, V>read()
+                .withConsumerConfigOverrides(getConsumerConfig())
+                .withOffsetConsumerConfigOverrides(getOffsetConsumerConfig())
+                .withConsumerFactoryFn(getConsumerFactoryFn())
+                .withKeyDeserializerProvider(getKeyDeserializerProvider())
+                .withValueDeserializerProvider(getValueDeserializerProvider())
+                .withExtractOutputTimestampFn(timestampFn);
+        if (isCommitOffsetsInFinalizeEnabled()) {
+          readTransform = readTransform.commitOffsets();
+        }
+
+        return input
+            .getPipeline()
+            .apply(Impulse.create())
+            .apply(
+                ParDo.of(
+                    new GenerateKafkaSourceDescription(
+                        readTransform.getKafkaSourceDescriptionSchema())))
+            .setCoder(RowCoder.of(readTransform.getKafkaSourceDescriptionSchema()))
+            .apply(readTransform)
+            .setCoder(KafkaRecordCoder.of(keyCoder, valueCoder));
+      }
+    }
+
+    /**
+     * A DoFn which generates {@link Row} with {@link KafkaSourceDescriptionSchemas#getSchema()}
+     * based on the configuration of {@link Read}.
+     */
+    @VisibleForTesting
+    class GenerateKafkaSourceDescription extends DoFn<byte[], Row> {
+      GenerateKafkaSourceDescription(Schema schema) {
+        this.kafkaSourceDescriptionSchema = schema;
+      }
+
+      private final Schema kafkaSourceDescriptionSchema;
+
+      private final Map<String, Object> consumerConfig = Read.this.getConsumerConfig();
```

Comment:
You should make the class static to prevent pulling in the parent transform during serialization. This would require passing forward arguments from the Read transform to the constructor. This will prevent accidentally capturing things we don't want captured.